### ISSUE TITLE
Cannot inherit from foam.parse.Parsers singleton

### ISSUE THREAD

**Issue by mdittmer** [foam-framework/foam2#287]
I wish to extend `foam.parse.Parsers` with more parsers. In particular, wish to get:

```
TokenParsers<Multiton<separator>> : Parsers<Singleton>
```

This doesn't work because the `Multiton` code will delegate to existing (custom `Singleton`) `create()` every time it attempts to create an instance.

Proposed solution: Separate plain class `foam.parse.Parsers` from singleton class `foam.parse.ParsersSingleton`. Existing code should refer to the latter. Extensions that add more parsers should extend the former.

**Comment by kgrgreer** [comment]
Can you refine instead?

**Comment by mdittmer** [comment]
I could. I have multiple parsers in the same namespace, some need tokenization and some don't. I suppose they could all have tokenization support, but some just don't use it. Does seem better than doing the same thing that we have for `Expressions` and `ExpressionsSingleton`?

**Comment by mdittmer** [comment]
Actually, revisiting the code, I'm pretty confident that I *cannot* refine to get the behaviour I want. If I refine `Parsers`, I'll be refining the singleton. However, `TokenParsers` is multiton over its `separator` property: each `TokenParsers` instance is bound to the separator used between tokens. This allows parser classes to do things like:

```js
foam.Function.withArgs(function(trepeat, tseq, tplus /* ... */) {
  /* ... */
}, TokenParsers.create({ separator: /* ... */ }), this);
```

Of course, this could be implemented differently; e.g., make every `t*(...)` parser get passed a separator parser as the first argument, but I feel this would be verbose. For the most part, tokenized parser definitions will res-use the same separator parser throughout.

**Comment by kgrgreer** [comment]
Singleton now detects if its been replaced and if it has then reverts to default behaviour. This should prevent it from interfering with Multitons.

**Comment by mdittmer** [comment]
Fixed in 7eef4801997c029ecee680366b18867e98631c87
